### PR TITLE
feat(shdict): multi-subsystems lua_shared_dict support

### PR DESCRIPTION
**Pull request by tzssangglass**

### CODE REVIEW

**Comment by tzssangglass** [comment]
Note: should revert `ci.yml` before merge to master

**Comment by tzssangglass** [comment]
build apisix-base with: https://github.com/api7/apisix-build-tools/pull/190

File: patch/README.md [hunk 1]
```diff
@@ -13,4 +13,15 @@ The `*-ngx_pipe_environ_on_mac` patches support the environ argument of the ngx.
...
+https://github.com/openresty/lua-resty-core/pull/276
+
+The `*-ngx_meta_lua_module` patches, `src/meta/*` and `t/meta/*` originally come from:
+https://github.com/openresty/meta-lua-nginx-module/pull/76, which is under BSD license.
+Copyright and License of this PR:
+This repository is licensed under the BSD license.
+
+Copyright (C) 2009-2017, by Xiaozhe Wang (chaoslawful) chaoslawful@gmail.com.
+
+Copyright (C) 2009-2019, by Yichun "agentzh" Zhang (章亦春) agentzh@gmail.com, OpenResty Inc.
```
Comment (spacewander) [inline patch/README.md:25]: We should remove this part as it is part of that license?

File: patch/1.19.9/lua-resty-core-ngx_meta_lua_module.patch [hunk 2]
```diff
@@ -0,0 +1,273 @@
+diff --git lib/resty/core/shdict.lua lib/resty/core/shdict.lua
```
Comment (spacewander) [inline patch/1.19.9/lua-resty-core-ngx_meta_lua_module.patch:1]: We should name the patch with its function `shared_shdict`

File: patch/1.19.9/lua-resty-core-ngx_meta_lua_module.patch [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
++
+ if not pcall(function () return C.free end) then
+     ffi.cdef[[
+ void free(void *ptr);
+@@ -179,7 +101,7 @@ local function check_zone(zone)
+         error("bad \"zone\" argument", 3)
+     end
+ 
+-    zone = ngx_lua_ffi_shdict_udata_to_zone(zone)
++    zone = C.ngx_meta_lua_ffi_shdict_udata_to_zone(zone)
```
Comment (spacewander) [inline patch/1.19.9/lua-resty-core-ngx_meta_lua_module.patch:170]: Better to use ngx_lua_ffi_shdict_udata_to_zone and update it with `ngx_lua_ffi_shdict_udata_to_zone = C.ngx_meta_lua_ffi_shdict_udata_to_zone`. And other similar places.

File: patch/README.md [hunk 4]
```diff
@@ -13,4 +13,15 @@ The `*-ngx_pipe_environ_on_mac` patches support the environ argument of the ngx.
 
 The `*-enable_keepalive` patches originally come from:
 https://github.com/openresty/lua-nginx-module/pull/1600
-https://github.com/openresty/lua-resty-core/pull/276
\ No newline at end of file
+https://github.com/openresty/lua-resty-core/pull/276
+
+The `*-ngx_meta_lua_module` patches, `src/meta/*` and `t/meta/*` originally come from:
+https://github.com/openresty/meta-lua-nginx-module/pull/76, which is under BSD license.
```
Comment (spacewander) [inline patch/README.md:19]: We should mention other PRs, like the one to lua-resty-core.

File: src/meta/config [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+ngx_module_name=ngx_meta_lua_module
+ngx_module_incs="$ngx_addon_dir"
+ngx_module_deps="$ngx_addon_dir/ddebug.h \
+                 $ngx_addon_dir/ngx_meta_lua_api.h \
+                 $ngx_addon_dir/ngx_meta_lua_module.h \
+                 $ngx_addon_dir/ngx_meta_lua_shdict.h"
+ngx_module_srcs="$ngx_addon_dir/ngx_meta_lua_module.c \
+                 $ngx_addon_dir/ngx_meta_lua_shdict.c"
+ngx_module_libs=
+ngx_module_link=ADDON
```
Comment (spacewander) [inline src/meta/config:11]: This variable is set by Nginx, not by us.
See https://www.nginx.com/resources/wiki/extending/converting/#complex-example

File: src/meta/config [hunk 6]
```diff
@@ -0,0 +1,13 @@
+ngx_module_type=CORE
+ngx_module_name=ngx_meta_lua_module
+ngx_module_incs="$ngx_addon_dir"
+ngx_module_deps="$ngx_addon_dir/ddebug.h \
+                 $ngx_addon_dir/ngx_meta_lua_api.h \
+                 $ngx_addon_dir/ngx_meta_lua_module.h \
+                 $ngx_addon_dir/ngx_meta_lua_shdict.h"
+ngx_module_srcs="$ngx_addon_dir/ngx_meta_lua_module.c \
+                 $ngx_addon_dir/ngx_meta_lua_shdict.c"
+ngx_module_libs=
```
Comment (spacewander) [inline src/meta/config:10]: We can remove the unset variable

File: patch/1.19.9/lua-resty-core-ngx_meta_lua_module.patch [hunk 7]
```diff
@@ -0,0 +1,273 @@
+diff --git lib/resty/core/shdict.lua lib/resty/core/shdict.lua
+index dedf12c..575cfb9 100644
+--- lib/resty/core/shdict.lua
++++ lib/resty/core/shdict.lua
+@@ -1,7 +1,7 @@
+ -- Copyright (C) Yichun Zhang (agentzh)
+ 
+ 
+-local ffi = require 'ffi'
++local ffi = require "ffi"
```
Comment (soulbird) [inline patch/1.19.9/lua-resty-core-ngx_meta_lua_module.patch:10]: This modification doesn't seem to make sense

Comment (tzssangglass) [reply patch/1.19.9/lua-resty-core-ngx_meta_lua_module.patch]: revert this

File: src/meta/config (hunk 6 above)
Comment (tzssangglass) [reply src/meta/config]: fix

File: src/meta/config (hunk 5 above)
Comment (tzssangglass) [reply src/meta/config]: fix

File: patch/README.md (hunk 4 above)
Comment (tzssangglass) [reply patch/README.md]: add commit link

File: patch/README.md (hunk 1 above)
Comment (tzssangglass) [reply patch/README.md]: remove this

File: patch/1.19.9/lua-resty-core-ngx_meta_lua_module.patch (hunk 2 above)
Comment (tzssangglass) [reply patch/1.19.9/lua-resty-core-ngx_meta_lua_module.patch]: changed

File: patch/1.19.9/lua-resty-core-ngx_meta_lua_module.patch (hunk 3 above)
Comment (tzssangglass) [reply patch/1.19.9/lua-resty-core-ngx_meta_lua_module.patch]: now it looks like 

``` lua
local ngx_lua_ffi_shdict_get
local ngx_lua_ffi_shdict_incr
local ngx_lua_ffi_shdict_store
local ngx_lua_ffi_shdict_flush_all
local ngx_lua_ffi_shdict_get_ttl
local ngx_lua_ffi_shdict_set_expire
local ngx_lua_ffi_shdict_capacity
local ngx_lua_ffi_shdict_free_space
local ngx_lua_ffi_shdict_udata_to_zone


ffi.cdef[[
int ngx_meta_lua_ffi_shdict_get(void *zone, const unsigned char *key,
    size_t key_len, int *value_type, unsigned char **str_value_buf,
    size_t *str_value_len, double *num_value, int *user_flags,
    int get_stale, int *is_stale, char **errmsg);

int ngx_meta_lua_ffi_shdict_incr(void *zone, const unsigned char *key,
    size_t key_len, double *value, char **err, int has_init,
    double init, long init_ttl, int *forcible);

int ngx_meta_lua_ffi_shdict_store(void *zone, int op,
    const unsigned char *key, size_t key_len, int value_type,
    const unsigned char *str_value_buf, size_t str_value_len,
    double num_value, long exptime, int user_flags, char **errmsg,
    int *forcible);

int ngx_meta_lua_ffi_shdict_flush_all(void *zone);

long ngx_meta_lua_ffi_shdict_get_ttl(void *zone,
    const unsigned char *key, size_t key_len);

int ngx_meta_lua_ffi_shdict_set_expire(void *zone,
    const unsigned char *key, size_t key_len, long exptime);

size_t ngx_meta_lua_ffi_shdict_capacity(void *zone);

void *ngx_meta_lua_ffi_shdict_udata_to_zone(void *zone_udata);
]]

if not pcall(function ()
    return C.ngx_meta_lua_ffi_shdict_free_space
end)
then
    ffi.cdef[[
size_t ngx_meta_lua_ffi_shdict_free_space(void *zone);
    ]]
end

pcall(function ()
    ngx_lua_ffi_shdict_get = C.ngx_meta_lua_ffi_shdict_get
    ngx_lua_ffi_shdict_incr = C.ngx_meta_lua_ffi_shdict_incr
    ngx_lua_ffi_shdict_store = C.ngx_meta_lua_ffi_shdict_store
    ngx_lua_ffi_shdict_flush_all = C.ngx_meta_lua_ffi_shdict_flush_all
    ngx_lua_ffi_shdict_get_ttl = C.ngx_meta_lua_ffi_shdict_get_ttl
    ngx_lua_ffi_shdict_set_expire = C.ngx_meta_lua_ffi_shdict_set_expire
    ngx_lua_ffi_shdict_capacity = C.ngx_meta_lua_ffi_shdict_capacity
    ngx_lua_ffi_shdict_free_space = C.ngx_meta_lua_ffi_shdict_free_space
    ngx_lua_ffi_shdict_udata_to_zone = C.ngx_meta_lua_ffi_shdict_udata_to_zone
end)

if not pcall(function () return C.free end) then
    ffi.cdef[[
void free(void *ptr);
    ]]
end
```

**Review by spacewander — APPROVED** [review]

**Review by soulbird — APPROVED** [review]

**Pull request merged**